build/lpc55pins/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+
+impl Pin {
+    fn get_port_pin(&self) -> (usize, usize) {
+        if self.pin > 32 {
+            panic!("Invalid pin {}", self.pin);
+        }
+
+        if self.port > 1 {
+            panic!("Invalid port {}", self.port);
+        }
+
+        (self.port, self.pin)
+    }
+}
+
+impl ToTokens for Pin {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let (port, pin) = self.get_port_pin();
+        let final_pin = format_ident!("PIO{}_{}", port, pin);
+        tokens.append_all(quote::quote! {
+            // Yes we want the trailing comma
+            Pin::#final_pin,
+        });
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct PinConfig {
+    pin: Pin,
+    alt: usize,
+    mode: Option<String>,
+    slew: Option<String>,
+    invert: Option<String>,
+    digimode: Option<String>,
+    opendrain: Option<String>,
+    direction: Option<String>,
+}
+
+impl PinConfig {
+    fn get_mode(&self) -> String {
+        match &self.mode {
+            None => "NoPull".to_string(),
+            Some(s) => s.to_string(),
+        }
+    }
+
+    fn get_slew(&self) -> String {
+        match &self.slew {
+            None => "Standard".to_string(),
+            Some(s) => s.to_string(),
+        }
+    }
+
+    fn get_invert(&self) -> String {
+        match &self.invert {
+            None => "Disable".to_string(),
+            Some(s) => s.to_string(),
+        }
+    }
+
+    fn get_digimode(&self) -> String {
+        match &self.digimode {
+            None => "Digital".to_string(),
+            Some(s) => s.to_string(),
+        }
+    }
+
+    fn get_opendrain(&self) -> String {
+        match &self.opendrain {
+            None => "Normal".to_string(),
+            Some(s) => s.to_string(),
+        }
+    }
+
+    fn get_alt(&self) -> usize {
+        if self.alt > 9 {
+            panic!("Invalid alt setting {}", self.alt);
+        }
+
+        self.alt
+    }
+}
+
+impl ToTokens for PinConfig {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let final_pin = self.pin.to_token_stream();
+        let alt_num = format_ident!("Alt{}", self.get_alt());
+        let mode = format_ident!("{}", self.get_mode());
+        let slew = format_ident!("{}", self.get_slew());
+        let invert = format_ident!("{}", self.get_invert());
+        let digimode = format_ident!("{}", self.get_digimode());
+        let od = format_ident!("{}", self.get_opendrain());
+
+        tokens.append_all(final_pin);
+        tokens.append_all(quote::quote! {
+            AltFn::#alt_num,
+            Mode::#mode,
+            Slew::#slew,
+            Invert::#invert,
+            Digimode::#digimode,
+            Opendrain::#od,
+        });
+    }
+}
+
+pub fn codegen(pins: Vec<PinConfig>) -> Result<()> {
+    let out_dir = std::env::var("OUT_DIR")?;
+    let dest_path = std::path::Path::new(&out_dir).join("pin_config.rs");
+    let mut file = std::fs::File::create(&dest_path)?;
+
+    writeln!(
```

Comment:
I _tried_ to convince `quote` to do this, but did not succeed due to the logic required in the loop!